calibration_assistant.cpp

```diff
@@ -0,0 +1,1247 @@
+            return;
+    }
+    m_pSlewBtn->Enable(false);
+    m_pCalibrateBtn->Enable(false);
+    if (!pFrame->CaptureActive)
+        pFrame->StartCapturing();
+    if (PhdController::Guide(GUIDEOPT_FORCE_RECAL, settle, wxRect(), &msg))
+    {
+        ShowStatus(_("Waiting for calibration to complete"));
+        m_monitoringCalibration = true;
+    }
+    else
+    {
+        ShowError(_("Calibration could not start - suspend any imaging automation apps"), false);
+        m_pSlewBtn->Enable(true);
+        m_pCalibrateBtn->Enable(true);
+    }
+}
+
+void CalibrationAssistant::ExplainResults()
+{
+    CalAssistExplanationDialog* dlg = new CalAssistExplanationDialog(m_lastResult);
+
+    dlg->ShowModal();
+}
+
+void CalibrationAssistant::OnTargetWest(wxCommandEvent& evt)
+{
+    if (m_pCurrEast->GetValue())
+        m_pWarning->SetLabelText(_("MERIDIAN FLIP!"));
+    else
+        m_pWarning->SetLabelText(wxEmptyString);
+}
+
+void CalibrationAssistant::OnTargetEast(wxCommandEvent& evt)
+{
+    if (m_pCurrWest->GetValue())
+        m_pWarning->SetLabelText("MERIDIAN FLIP!");
+    else
+        m_pWarning->SetLabelText(wxEmptyString);
+}
+
+void CalibrationAssistant::OnCancel(wxCommandEvent& evt)
+{
+    wxDialog::Destroy();
+}
+
+void CalibrationAssistant::OnExplain(wxCommandEvent& evt)
+{
+    //m_lastResult = "Steps, Rates, Orthogonality, Sky location, Incomplete";
+    ExplainResults();
+}
+void CalibrationAssistant::OnClose(wxCloseEvent& evt)
+{
+    wxDialog::Destroy();
+}
+
+void CalibrationAssistant::OnRestore(wxCommandEvent& evt)
+{
+    InitializeUI(true);
+}
+
+void CalibrationAssistant::OnLoadCustom(wxCommandEvent& evt)
+{
+    InitializeUI(false);
+}
+
+void CalibrationAssistant::OnCustom(wxCommandEvent& evt)
+{
+    int ha = m_pTargetOffset->GetValue();
+    int dec = m_pTargetDec->GetValue();
+    bool sideEast = m_pTargetEast->GetValue() == 1;
+    if (sideEast)
+        ha = -ha;
+    CalCustomDialog custDlg(this, ha, dec);
+    custDlg.ShowModal();            // Dialog handles the UI updates
+}
+
+CalCustomDialog::CalCustomDialog(CalibrationAssistant* Parent, int DefaultHA, int DefaultDec)
+    : wxDialog(pFrame, wxID_ANY, _("Save Customized Calibration Position"),
+    wxDefaultPosition, wxSize(474, -1), wxCAPTION | wxCLOSE_BOX)
+{
+    m_Parent = Parent;
+    wxStaticBoxSizer* tgtSizer = new wxStaticBoxSizer(wxVERTICAL, this, _("Target Position"));
+    wxFlexGridSizer* targetPosSizer = new wxFlexGridSizer(1, 5, 5, 15);
+
+    int spinnerWidth = StringWidth(this, "0000");
+    m_pTargetDec = NewSpinnerInt(this, wxSize(spinnerWidth, -1), DefaultDec, -50, 50, 5, _("Target declination for slew, as close to Dec = 0 as possible for your location"));
+    AddTableEntryPair(this, targetPosSizer, _("Declination"), m_pTargetDec);
+    m_pTargetOffset = NewSpinnerInt(this, wxSize(spinnerWidth, -1), abs(DefaultHA), 5, 50, 5, _("Target offset from central meridian, in degrees; east or west based on 'Pointing' buttons"));

```

Comment:
should we ask for this value in hours then convert to degrees internally?  probably most planetarium and imaging apps will report hour angle in hours not degrees?
we could even ask for the target RA and compute hour angle ourselves, right? (or maybe not, that's assuming they want to image the target _now_)

Reply:
A lot of this is for users who don't understand the sky coordinate system and have no concept of why right ascension and declination are in different units.  Someone who is dealing with sky obstructions can probably have an intuitive notion of x degrees from the "center" but they aren't going to translate that into x/15 hours.  I don't think a planetarium app has any meaning here, it's PHD2 that's going to slew the scope.  We're not going to point at a particular target, we're going to point "over there".  In my view, changing the hour angle will be an unusual case and mostly done to deal with obstruction problems (balcony location, trees, neighbor's house...).

Reply:
ok.    I think what threw me off was "target" -- when I saw that I thought "imaging target" -- but this is really the calibration target location.  Should we avoid the word target to avoid mixing it up with imaging target?  Maybe call it "Calibration position" or "Calibration location" or "Calibration pointing position"?

Reply:
Ok, yes, "target" is probably a loaded term.  I will try "Calibration location".  This may also help discourage the notion that you need to choosing a specific star for calibration.